ServiceLoaderTest.testSkipHookLoadedByDifferentClassloader failure

*master* (commit 4525ce5ac463caa640d330ceafbef750448256e9)

Failing constantly only on IBM JDK 8: https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-IbmJDK8/lastCompletedBuild/testReport/com.hazelcast.util/ServiceLoaderTest/testSkipHookLoadedByDifferentClassloader/

Other failures (with same stacktrace): 
ServiceLoaderTest.testSkipHooksWithImplementingTheExpectedInterfaceButLoadedByDifferentClassloader
ServiceLoaderTest.testFailFastWhenHookDoesNotImplementExpectedInteface

Stacktrace:
```
java.lang.IllegalStateException: Cannot access final java.lang.Package java.lang.ClassLoader.getDefinedPackage(java.lang.String)
	at net.bytebuddy.dynamic.loading.ByteArrayClassLoader$PackageLookupStrategy$ForJava9CapableVm.apply(ByteArrayClassLoader.java:561)
	at net.bytebuddy.dynamic.loading.ByteArrayClassLoader$ClassDefinitionAction.run(ByteArrayClassLoader.java:439)
	at net.bytebuddy.dynamic.loading.ByteArrayClassLoader$ClassDefinitionAction.run(ByteArrayClassLoader.java:409)
	at java.security.AccessController.doPrivileged(AccessController.java:666)
	at net.bytebuddy.dynamic.loading.ByteArrayClassLoader.findClass(ByteArrayClassLoader.java:262)
	at java.lang.ClassLoader.loadClassHelper(ClassLoader.java:925)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:870)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:853)
	at java.lang.Class.forNameImpl(Native Method)
	at java.lang.Class.forName(Class.java:402)
	at net.bytebuddy.dynamic.loading.ByteArrayClassLoader.load(ByteArrayClassLoader.java:220)
	at net.bytebuddy.dynamic.loading.ClassLoadingStrategy$Default$WrappingDispatcher.load(ClassLoadingStrategy.java:285)
	at net.bytebuddy.dynamic.loading.ClassLoadingStrategy$Default.load(ClassLoadingStrategy.java:120)
	at net.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:79)
	at net.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:4376)
	at net.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:4362)
	at com.hazelcast.util.ServiceLoaderTest.newInterface(ServiceLoaderTest.java:243)
	at com.hazelcast.util.ServiceLoaderTest.testSkipHookLoadedByDifferentClassloader(ServiceLoaderTest.java:176)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:811)
Caused by: java.lang.IllegalAccessException: Class net.bytebuddy.dynamic.loading.ByteArrayClassLoader$PackageLookupStrategy$ForJava9CapableVm can not access a member of class java.lang.ClassLoader with modifiers "final"
	at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:114)
	at java.lang.reflect.AccessibleObject.slowCheckMemberAccess(AccessibleObject.java:399)
	at java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:391)
	at java.lang.reflect.Method.invoke(Method.java:501)
	at net.bytebuddy.dynamic.loading.ByteArrayClassLoader$PackageLookupStrategy$ForJava9CapableVm.apply(ByteArrayClassLoader.java:559)
	... 29 more
```

The failure is constant, actually much earlier, but because of Jenkins jobs migration, I'm reporting it now with quite a big delay, sorry about that.

--- a/hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastProxyFactory.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/starter/HazelcastProxyFactory.java
@@ -436,7 +436,7 @@ public class HazelcastProxyFactory {
         public static final AllAsPublicConstructorStrategy INSTANCE = new AllAsPublicConstructorStrategy();
 
         @Override
-        public MethodRegistry inject(MethodRegistry methodRegistry) {
+        public MethodRegistry inject(TypeDescription instrumentedType, MethodRegistry methodRegistry) {
             return methodRegistry.append(new LatentMatcher.Resolved<MethodDescription>(isConstructor()),
                     new MethodRegistry.Handler.ForImplementation(SuperMethodCall.INSTANCE),
                     MethodAttributeAppender.NoOp.INSTANCE,

--- a/hazelcast/src/test/java/com/hazelcast/util/ServiceLoaderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/util/ServiceLoaderTest.java
@@ -96,10 +96,10 @@ public class ServiceLoaderTest extends HazelcastTestSupport {
 
     @Test
     public void testHookDeduplication() {
-        ClassLoader parentClassloader = PortableHook.class.getClassLoader();
-
         Class<?> hook = newClassImplementingInterface("com.hazelcast.internal.serialization.SomeHook",
-                PortableHook.class, parentClassloader);
+                PortableHook.class, PortableHook.class.getClassLoader());
+
+        ClassLoader parentClassloader = hook.getClassLoader();
 
         //child classloader delegating everything to its parent
         URLClassLoader childClassloader = new URLClassLoader(new URL[]{}, parentClassloader);

--- a/pom.xml
+++ b/pom.xml
@@ -85,7 +85,7 @@
         <hamcrest.version>1.3</hamcrest.version>
         <mockito.version>2.19.0</mockito.version>
         <powermock.version>2.0.0-beta.5</powermock.version>
-        <bytebuddy.version>1.6.12</bytebuddy.version>
+        <bytebuddy.version>1.8.17</bytebuddy.version>
         <reflections.version>0.9.10</reflections.version>
         <jmh.version>1.16</jmh.version>
         <commons-lang3.version>3.4</commons-lang3.version>

